### ISSUE TITLE
Memory blows up using CNN model with multiple clients

### ISSUE THREAD

**Issue by JJJJJamie** [tensorflow/federated#751]
I tried to implement the tutorial of federated image classification with a CNN model, but it turns out that while I set the client number to 100, each round of training takes up about 16 Gb of memory and it stacks on after each round.  I am not sure that is there any issue with my code or is it because that after each round the weight tensors are not deleted.  I would appreciate it so much if anyone could help.  Below is my python code:
  
`from __future__ import absolute_import
from six.moves import range
import tensorflow as tf
from tensorflow.python.keras.optimizer_v2 import gradient_descent
import tensorflow_federated as tff
import collections
import time
import random

from tensorflow_federated.python.examples import mnist
tf.compat.v1.enable_v2_behavior()

NUM_CLIENTS = 100
NUM_EPOCHS = 5
BATCH_SIZE = 10
ROUNDS = 100
LEARNING_RATE = 0.02

tff.framework.set_default_executor(
    tff.framework.create_local_executor(NUM_CLIENTS))

source, _ = tff.simulation.datasets.emnist.load_data()

def client_data(n):
  dataset = source.create_tf_dataset_for_client(source.client_ids[n])
  return mnist.keras_dataset_from_emnist(dataset).repeat(NUM_EPOCHS).batch(BATCH_SIZE)

train_data = [client_data(n) for n in range(NUM_CLIENTS)]

sample_batch = tf.nest.map_structure(
    lambda x: x.numpy(), iter(train_data[0]).next())

def CNN_compiled():
    model = tf.keras.models.Sequential([
        tf.keras.layers.Reshape((28, 28, 1), input_shape=(784,)),
        tf.keras.layers.Conv2D(32, kernel_size=(5, 5), activation="relu", padding="same", strides=1),
        tf.keras.layers.MaxPooling2D(pool_size=2, strides=2, padding='valid'),
        tf.keras.layers.Conv2D(64, kernel_size=(5, 5), activation="relu", padding="same", strides=1),
        tf.keras.layers.MaxPooling2D(pool_size=2, strides=2, padding='valid'),
        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(512, activation="relu"),
        tf.keras.layers.Dense(10, activation="softmax"),
    ])
    def loss_fn(y_true, y_pred):
        return tf.reduce_mean(tf.keras.losses.sparse_categorical_crossentropy(
            y_true, y_pred))
    model.compile(
        loss=loss_fn,
        optimizer=gradient_descent.SGD(LEARNING_RATE),  # decay
        metrics=[tf.keras.metrics.SparseCategoricalAccuracy()])
    return model

def model_fn():
  keras_model = CNN_compiled()
  return tff.learning.from_compiled_keras_model(keras_model, sample_batch)

trainer = tff.learning.build_federated_averaging_process(model_fn)
state = trainer.initialize()
for _ in range(ROUNDS):
    now = time.time()
    state, metrics = trainer.next(state, train_data)
    time_spent = time.time() - now
    print (metrics, time_spent)`

**Comment by jiachangliu** [comment]
Hello, I encountered the same problem, and the memory accumulated after each round of training and exploded after some time. Has the team found a cause or solution to this problem? @jkr26 @ZacharyGarrett

**Comment by jkr26** [comment]
Given [this recent commit](https://github.com/tensorflow/federated/commit/5299226ee7c736c46d1d3def8c6726214f6f9cb0), this issue should be mitigated. Can those seeing this problem confirm?

**Comment by jkr26** [comment]
Closing since this should be no longer an issue.
